src/modules/mavlink/mavlink_receiver.cpp

```diff
@@ -1382,31 +1397,47 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
 					}
 
 					if (!offboard_control_mode.ignore_thrust) { // dont't overwrite thrust if it's invalid
-						// Fill correct field by checking frametype
-						// TODO: add as needed
-						switch (_mavlink->get_system_type()) {
-						case MAV_TYPE_GENERIC:
-							break;
-
-						case MAV_TYPE_FIXED_WING:
-							att_sp.thrust_body[0] = set_attitude_target.thrust;
-							break;
+						if (!_is_vtol) {
+							// Fill correct field by checking frametype
+							// TODO: add as needed
+							switch (_mavlink->get_system_type()) {
+							case MAV_TYPE_GENERIC:
+								break;
+
+							case MAV_TYPE_FIXED_WING:
+								att_sp.thrust_body[0] = set_attitude_target.thrust;
+								break;
+
+							case MAV_TYPE_QUADROTOR:
+							case MAV_TYPE_HEXAROTOR:
+							case MAV_TYPE_OCTOROTOR:
+							case MAV_TYPE_TRICOPTER:
+							case MAV_TYPE_HELICOPTER:
+								att_sp.thrust_body[2] = -set_attitude_target.thrust;
+								break;
+
+							case MAV_TYPE_GROUND_ROVER:
+								att_sp.thrust_body[0] = set_attitude_target.thrust;
+								break;
+							}
+
+							_att_sp_pub.publish(att_sp);
 
-						case MAV_TYPE_QUADROTOR:
-						case MAV_TYPE_HEXAROTOR:
-						case MAV_TYPE_OCTOROTOR:
-						case MAV_TYPE_TRICOPTER:
-						case MAV_TYPE_HELICOPTER:
-							att_sp.thrust_body[2] = -set_attitude_target.thrust;
-							break;
-
-						case MAV_TYPE_GROUND_ROVER:
-							att_sp.thrust_body[0] = set_attitude_target.thrust;
-							break;
+						} else {
+							if (_vtol_vehicle_status_sub.updated()) {
+								_vtol_vehicle_status_sub.copy(&_vtol_vehicle_status);
+							}
+
+							if (_vtol_vehicle_status.vtol_in_rw_mode) {
```

Comment:
Wouldn't this be better being handled in `mc_att_control` or `fw_att_control` rather than being handled in the mavlink receiver? This creates a dependency of mavlink_receiver to mc and fw attitude controllers

Reply:
@Jaeyoung-Lim How would you handle this in the controllers? The controllers receive a 3D thrust setpoint, it would be incorrect for the FW controller to use the Z component of the thrust setpoint, nor for MC controller to use the X component of the thrust setpoint.

The real alternative is to extend mavlink so that the offboard sender can properly handle this.

Reply:
Ah, that makes sense

Reply:
Hm. It would be really better.

But if I uderstand correctly at this time its too big changes in architecture:

- `fw_virtual_attitude_setpoint` used at least in [FixedwingAttitudeControl](https://github.com/PX4/Firmware/blob/26364d44c91672d101349ff5825358017f7db3df/src/modules/fw_att_control/FixedwingAttitudeControl.cpp), in [FixedwingPositionControl](https://github.com/PX4/Firmware/blob/26364d44c91672d101349ff5825358017f7db3df/src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp) and in [vtol_att_control_main](https://github.com/PX4/Firmware/blob/26364d44c91672d101349ff5825358017f7db3df/src/modules/vtol_att_control/vtol_att_control_main.h).

- `mc_virtual_attitude_setpoint` used at least in [FixedwingPositionControl](https://github.com/PX4/Firmware/blob/26364d44c91672d101349ff5825358017f7db3df/src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp), [mc_pos_control_main](https://github.com/PX4/Firmware/blob/e3bbac0447b96e483db2227c0b900ab64a4e733a/src/modules/mc_pos_control/mc_pos_control_main.cpp).

Or we can add new uORB topic for use it in switch [like that](https://github.com/PX4/Firmware/blob/f885d2274a563278b5ca501e4c19365ad0cd790d/src/modules/vtol_att_control/vtol_att_control_main.cpp#L382) to check and publish in proper `mc_virtual_attitude_setpoint` or `fw_virtual_attitude_setpoint`, but it is also does not a small work.